### PR TITLE
Add array-fast path in Enumerable.Min/Max(int/long)

### PR DESCRIPTION
**Pull request by stephentoub**
In looking through usage data, it's pretty common to call Enumerable.Min/Max with an array.  At the cost of a type check when the source isn't an array, we can speed up the cases where it is, at a minimum by avoiding interface dispatch and in the best case by vectorizing.  I've done so for int/long.  We could look at doing so for float/double as well, but I wasn't sure about the impact of the NaN-related logic on vectorization.

| Method |         Toolchain | Length | Array |         Mean |      Error | Ratio |
|------- |------------------ |------- |------ |-------------:|-----------:|------:|
|    Min | \main\corerun.exe |      1 | False |    31.444 ns |  0.6275 ns |  1.00 |
|    Min |   \pr\corerun.exe |      1 | False |    31.211 ns |  0.3926 ns |  0.99 |
|        |                   |        |       |              |            |       |
|    Min | \main\corerun.exe |      1 |  True |    14.049 ns |  0.0917 ns |  1.00 |
|    Min |   \pr\corerun.exe |      1 |  True |     1.912 ns |  0.0116 ns |  0.14 |
|        |                   |        |       |              |            |       |
|    Min | \main\corerun.exe |      4 | False |    44.083 ns |  0.5091 ns |  1.00 |
|    Min |   \pr\corerun.exe |      4 | False |    44.900 ns |  0.4524 ns |  1.02 |
|        |                   |        |       |              |            |       |
|    Min | \main\corerun.exe |      4 |  True |    24.681 ns |  0.1511 ns |  1.00 |
|    Min |   \pr\corerun.exe |      4 |  True |     3.553 ns |  0.0647 ns |  0.14 |
|        |                   |        |       |              |            |       |
|    Min | \main\corerun.exe |     16 | False |   108.921 ns |  1.0189 ns |  1.00 |
|    Min |   \pr\corerun.exe |     16 | False |   112.215 ns |  0.5962 ns |  1.03 |
|        |                   |        |       |              |            |       |
|    Min | \main\corerun.exe |     16 |  True |    74.743 ns |  0.4251 ns |  1.00 |
|    Min |   \pr\corerun.exe |     16 |  True |     4.557 ns |  0.0403 ns |  0.06 |
|        |                   |        |       |              |            |       |
|    Min | \main\corerun.exe |     64 | False |   333.232 ns |  1.3208 ns |  1.00 |
|    Min |   \pr\corerun.exe |     64 | False |   331.662 ns |  6.6332 ns |  0.99 |
|        |                   |        |       |              |            |       |
|    Min | \main\corerun.exe |     64 |  True |   251.663 ns |  5.0050 ns |  1.00 |
|    Min |   \pr\corerun.exe |     64 |  True |     7.923 ns |  0.0847 ns |  0.03 |
|        |                   |        |       |              |            |       |
|    Min | \main\corerun.exe |    256 | False | 1,275.711 ns | 25.5032 ns |  1.00 |
|    Min |   \pr\corerun.exe |    256 | False | 1,220.111 ns | 24.3144 ns |  0.96 |
|        |                   |        |       |              |            |       |
|    Min | \main\corerun.exe |    256 |  True |   932.077 ns |  2.8967 ns |  1.00 |
|    Min |   \pr\corerun.exe |    256 |  True |    21.124 ns |  0.1536 ns |  0.02 |
|        |                   |        |       |              |            |       |
|    Min | \main\corerun.exe |   1024 | False | 4,951.882 ns | 92.7338 ns |  1.00 |
|    Min |   \pr\corerun.exe |   1024 | False | 4,673.227 ns | 41.3033 ns |  0.94 |
|        |                   |        |       |              |            |       |
|    Min | \main\corerun.exe |   1024 |  True | 3,753.438 ns | 50.7476 ns |  1.00 |
|    Min |   \pr\corerun.exe |   1024 |  True |    86.373 ns |  0.3499 ns |  0.02 |

```C#
using System;
using System.Collections.Generic;
using System.Linq;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

public class Program
{
    public static void Main(string[] args) => BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly).Run(args);

    [Params(1, 4, 16, 64, 256, 1024)]
    public int Length { get; set; }

    [Params(false, true)]
    public bool Array { get; set; }

    private IEnumerable<int> _source;

    [GlobalSetup]
    public void Setup()
    {
        _source = Enumerable.Range(1, Length).Reverse();
        if (Array)
        {
            _source = _source.ToArray();
        }
    }

    [Benchmark]
    public int Min() => _source.Min();
}
```

### CODE REVIEW

File: src/libraries/System.Linq/src/System/Linq/Max.cs
Comment (tannergooding) [inline src/libraries/System.Linq/src/System/Linq/Max.cs:45]: So what I'd really like to see here is for many of these methods to be available on `Span<T>` and for us to forward this to the Span implementation.

Exposing new public APIs is of course something that needs to goto API proposal, but I think it would be better long term, would also allow more logic sharing for various helper APIs, and would centralize most of the "array like SIMD algorithms" to the same place in the BCL.

File: src/libraries/System.Linq/src/System/Linq/Max.cs
Comment (tannergooding) [inline src/libraries/System.Linq/src/System/Linq/Max.cs:54]: Do we know how big the average input here is? This is going to give us different perf characteristics between Arm64 and x64 because the former will trigger for `8+` element arrays (`32-bytes`) while the latter will only trigger for `16+` element arrays (`64-bytes`).

I'd expect we want similar handling in both and while manually unrolling may pipeline well on x64, that may not be the case on Arm64 or all x64 based chipsets.

Generally the using `Vector256` is beneficial for very large inputs where you can saturate the decoder and memory pipeline; but are less beneficial for "small" inputs (particularly on older hardware where cache line splits and throughput may be more limited).

File: src/libraries/System.Linq/src/System/Linq/Max.cs
Comment (stephentoub) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:45]: I'd be fine subsequently seeing such methods on MemoryExtensions and deleting the code from here to delegate there instead.  Someone would need to do the work to figure out what that set of methods should be.  I also don't think it precludes adding this now and then delegating later if/when the methods exist.

Comment (tannergooding) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:45]: Right, I just wanted to call out what I'd like to see here and I think just about any method we think is worth special-casing for `array` in `LINQ` is applicable.

Sum/Min/Max are the "most obvious ones"

File: src/libraries/System.Linq/src/System/Linq/Max.cs
Comment (stephentoub) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:54]: > Do we know how big the average input here is? 

I'm not sure.

Are you suggesting a default stance should be to use `Vector128<T>` instead of `Vector<T>` unless you know big inputs are common?  The tradeoff you're describing is between potentially 2x throughput on larger sizes (if you use `Vector<T>` and end up employing 256-bit operations instead of 128-bit operations) vs vectorizing smaller sizes (if you use `Vector<T>` and can't vectorize the processing of 200 bits but could have a little with `Vector128<T>`)?

Comment (stephentoub) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:54]: (This is why I opened https://github.com/dotnet/runtime/issues/64409.  We need really clear guidance on when to use which.)

Comment (tannergooding) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:54]: > Are you suggesting a default stance should be to use Vector128<T> instead of Vector<T> unless you know big inputs are common? The tradeoff you're describing is between potentially 2x throughput on larger sizes (if you use Vector<T> and end up employing 256-bit operations instead of 128-bit operations) vs vectorizing smaller sizes (if you use Vector<T> and can't vectorize the processing of 200 bits but could have a little with Vector128<T>)?

Yes. There is a lot of nuance here such as `Arm64` not having `Vector256<T>` and it often having different pipelining characteristics. Even when just considering x64, there are various things to consider as well such as the expected target hardware as modern (Skylake and Zen2 or later) tends to perform really well with V256; where-as older (Haswell/Broadwell or Zen1) tends to behave worse and may incur downclocking, may not support dual dispatch, and may even emulate 256-bit support as 2x128-bit ops.

I'd generally expect that we do something like:
* `Vector128<T>` is the default choice and is used for practically any input that is at least `Count` long
  * This is the path that will light up basically "everywhere" (Arm64, x64, x86; eventually WASM, etc)
  * It may or may not be worth it to specialize cases where we are doing less 4-7 loop iterations due to the default branch prediction behavior
* `Vector256<T>` is used for cases with at least 128-bytes. This is also the cutoff of many memcpy algorithms before they start considering `non-temporal` or other store sizes
  * This path will only light up on x86/x64 and can be impacted by things like alignment, whether its going to split the cache every other read/write, the instructions being used, and the underlying data size

There are likely cases where microbenches or modern hardware (roughly Skylake and Zen2 or newer) will show `Vector256<T>` is better. However, I expect that a good bit of hardware (for both servers and consumers) still in the wild isn't this and so we'll need to take that into consideration. Likewise, the newer `Alder Lake` CPUs also bring in considerations where the `Vector256<T>` support is slower on the energy efficient cores (`vpmaxud` is 1 latency and 0.5 throughput for V128 and V256 on the power cores; but is `1` latency, `0.33` throughput on the energy cores for V128 and `2` latency, `0.67` throughput for V256)

Comment (tannergooding) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:54]: Noting that I wouldn't block this PR on that doc being written up. I think this largely matches what we're already doing for other cases.

Just that I believe this is roughly what the writeup will follow based on my own knowledge/experience.

**Review by tannergooding — APPROVED** [review]
LGTM and matches what we are doing for other similar code already.

We probably want to look again later after I do the doc writeup.

File: src/libraries/System.Linq/src/System/Linq/Max.cs
Comment (stephentoub) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:54]: Thanks. I take what you wrote above to mean essentially that we'd need to revise pretty much every existing vectorized implementation we have. I don't believe most or any set a threshold of 128 bytes in order to use 256-bit operations, and using `Vector<T>` is incompatible with what you wrote.

This also makes me nervous about the path we're on. You've essentially said to not use `Vector<T>`. Should we deprecate it? Make it behave like `Vector128`? Something else?

I'm very much looking forward to the compleye guidance and to us ensuring all of our implementations conform.

**Comment by stephentoub** [comment]
thank you for reviewing and for the pre-look at your thinking :-)

File: src/libraries/System.Linq/src/System/Linq/Max.cs
Comment (tannergooding) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:54]: >  I don't believe most or any set a threshold of 128 bytes in order to use 256-bit operations, and using Vector<T> is incompatible with what you wrote.

Right. We basically have two categories of vectorized code today:
1. Code that was written before hardware intrinsics and uses Vector<T>
2. Code that was written because of new functionality hardware intrinsics enabled and which is largely just considering Skylake based CPUs

I believe the introduction of `Arm64` as a major platform, the recent change by Intel to use heterogenous CPUs with both Power and Efficiency cores, the continued prevalence of Haswell/Broadwell CPUs in the low/mid end cloud scenarios, the potential for future WASM SIMD support, and the general need for us to support a range of platforms means that we should adjust the patterns we're doing here to ensure that we run well in all the scenarios here, even if it leaves some perf on the table in the "large input" scenarios for the absolute latest hardware.

> This also makes me nervous about the path we're on. You've essentially said to not use Vector<T>. Should we deprecate it? Make it behave like Vector128 on some additional platforms? Something else?

This is a fairly complex topic and part of it comes down to us being the BCL and having the diverse set of needs and requirements detailed above. This may not be applicable to many types of user code and `Vector<T>` is still a very good option if you know the data is likely to always be "big" (at least several KB in size).

The longer explanation is that `Vector<T>` still generates good code and is the only way to provide SIMD acceleration if you are needing to multitarget .NET Standard/Framework. It also has some benefits in being variable sized, such as automatically scaling to the hardware. This functionality is particularly beneficial for cases that are known to be "big data" and where small inputs are basically never expected. This can also be seen in Intel introducing AVX-512 and ARM introducing SVE/SVE2 (the latter being explicitly "variable" sized rather than an additional "fixed-width" vector option).

That being said, this benefit is also its biggest drawback. It being variable sized means that certain functionality is extremely difficult to represent. You can't for example have an equivalent of `Vector128.Create(1.0f, 2.0f, 3.0f, 4.0f)`, where you create a vector from all constant inputs, as the `Count` differs from platform to platform. Because of this, you also cannot trivially represent something like ` Shuffle` operation and you cannot trivially provide a way to extract the most significant bit of each element. This restricts the ability to use `Vector<T>` in a number of scenarios that we've had in the BCL. Likewise, you can't use `Vector<T>` to accelerate other "fixed" sized operations that may be common in certain graphics or multimedia based scenarios (3D audio, 4x4 matrices, 2/3/4 dimensional vectors, etc). The finally "issue" with being variable size is it becomes hard to know when something will or won't be accelerated. For example, you'd get acceleration for 16+ byte inputs on Arm64; but only for 32+ byte inputs on modern x86/x64. While older x86/64 (~pre 2013) would get acceleration for 16+ byte inputs. Depending on what is or is not common for the API this may be good or bad.

Now some of these issues could be fixed/resolved. There are opportunities for us to provide additional helper APIs to cover things like "I only have less than `Vector<T>.Count` elements of data left" (this also applies to `Vector64/128/256<T>`) and while having a fixed width is beneficial in many cases it and the additional APIs having a fixed width enables are not required in a lot of scenarios.

So I think the general guidance is summed up as:
* A Vector<T> path is not recommended unless you are multi-targeting .NET Standard or expect the data to always be over `big_data_threshold` bytes
  * I believe `big_data_threshold` is likely 2, 4, or 8kb. This represent `0.5`, `1`, and `2` "typical" pages of memory; respectively
  * 2kb is also the "largest" SVE vector size that can be supported on Arm64 and I expect the largest we'd reasonably encounter
* A Vector128<T> path should always be provided as it allows consistent SIMD acceleration on all platforms we care about
* Consider a Vector256<T> path if the data is expected to commonly be over `v256_threshold` bytes
  * I believe `v256_threshold` bytes is likely 64, 128 or 256 bytes. This represents `1`, `2`, or `4` "typical" cache lines; respectively
  * These are common thresholds hit in C algorithms, particularly in things like `memcpy` that have undergone years of fine tuning

There are likely some additional considerations to be had depending on how important bits of perf are that can make the algorithms more or less complex; some of which may be solvable by helper APIs that allow more trivially dealing with things like leading/trailing elements when you don't have a full `VectorXXX<T>.Count` elements remaining.

Comment (tannergooding) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:54]: A lot of this is based on my own predictions/experience here and is influenced based on what the various Optimization Manuals from AMD/Intel/Arm recommend as well as from known information available about the actual micro-architecture pipelines and characteristics of CPUs from ~2013 onwards (this is essentially when AVX2 on x86/x64 and ARMv8-A, which is the first Arm64 version, were introduced and so is a good "baseline" to set ourselves to).

Its possible that we have other needs, data, or driving scenarios that modify this a bit and I'm happy to take those into account in the doc I write up, as appropriate.

File: src/libraries/System.Linq/tests/Shuffler.cs
Comment (davidfowl) [inline src/libraries/System.Linq/tests/Shuffler.cs:17]: ❤️

File: src/libraries/System.Linq/src/System/Linq/Max.cs
Comment (eiriktsarpalis) [inline src/libraries/System.Linq/src/System/Linq/Max.cs:18]: Why not `source is int[] array`?

Comment (stephentoub) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:18]: It's measurably more expensive here. This check is as close to free as possible, a simple compare against the type token, whereas the is check ends up incurring a non-inlined call to a helper function, as technically the is can be satisfied by other types, e.g. uint[]. It adds enough overhead that it was showing up at a 10-20% regression on smaller non-inputs. The overhead with the explicit check is small enough that we might consider adding array handling in other such sink methods, even if not vectorizable.

File: src/libraries/System.Linq/tests/Shuffler.cs
Comment (eiriktsarpalis) [inline src/libraries/System.Linq/tests/Shuffler.cs:16]: Maybe we should use a seeded `Random` instance here for deterministic test inputs?

**Review by eiriktsarpalis — APPROVED** [review]

File: src/libraries/System.Linq/tests/Shuffler.cs
Comment (stephentoub) [reply src/libraries/System.Linq/tests/Shuffler.cs:16]: Sure.  I'll do that in a separate PR.

File: src/libraries/System.Linq/src/System/Linq/Max.cs
Comment (stephentoub) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:54]: > Its possible that we have other needs, data, or driving scenarios that modify this a bit and I'm happy to take those into account in the doc I write up, as appropriate.

Thanks for all your thoughts.

My top-level concern here is that we're going to be pushing people to a more complicated place that leads to most developers at least duplicating the relevant code paths and also needing to factor in complicated information that will leave them scratching their heads.  You say essentially the default position should be to add a `Vector128<T>` path and then optionally add a `Vector256<T>` path if you might have at least 128 bytes of input, but I expect (and see based on our own usage) that anywhere we use `Vector<T>` today we invariably expect an unbounded amount of data, and we would not be willing to leave 2x throughput on the table.  I don't think guidance then that pushes people towards such more complicated code is the right answer, even if it is the answer that eeks out an additional X% perf in some cases.

I'd prefer to see our guidance be along the lines of:
- Default to using `Vector<T>` if the API surface suits your needs (and do our best to add whatever APIs we need to ensure that it can).  You write your vectorized code once, and the system optimizes it well for whatever architecture your system supports.  This leads to smaller IL, smaller C#, more maintainable code, etc.
- If the API surface area of `Vector<T>` doesn't support what you need, or if it's critical for your scenario to optimize for inputs >= 128 bits but < 256 bits, then consider using `Vector128<T>` instead, and assuming you also want to optimize for larger inputs as well, add a second path for `Vector256<T>`.  If you do add a `Vector256<T>` path, you should fall back to using the `Vector128<T>` path in the following scenarios with the following general structure... followed by whatever that is :)
- If the rare vectorization scenario where the API surface area of `Vector128<T>` and `Vector256<T>` do not meet your needs, you do have access to the direct intrinsics APIs, but direct usage should be reserved only for the most fine-tuned cases where you performance testing demonstrates a true benefit, etc. etc.

Comment (tannergooding) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:54]: I think it might be good/easier for us to have a 15-30m sync up on Teams to discuss this a bit more in depth.

There is a lot of nuance here including how `Vector<T>` impacts the common cases and makes just supporting `Vector128<T>` a better option, IMO.

Comment (stephentoub) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:54]: > I think it might be good/easier for us to have a 15-30m sync up on Teams to discuss this a bit more in depth.

Ok

> There is a lot of nuance here including how Vector<T> impacts the common cases and makes just supporting Vector128<T> a better option, IMO.

I understand there's a lot of nuance.  And I don't see how "just supporting Vector128<T>" will actually work in practice; unless you know all of your inputs are small, which is rare, if you're reaching for Vector* you care about throughput and you're not going to just leave 2x throughput on the table.  What percentage of our existing vectorized implementations could have provided a 256-bit code path but stopped at a 128-bit code path?

Comment (tannergooding) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:54]: One of the nuances here is that `256-bit` is x86/x64 only. It doesn't apply to any of the other architectures/platforms we support (Arm, Arm64, WASM, etc) and it won't apply to most theoretical platforms we could support.

It's effectively a specialization only for x86/x64 CPUs that are from 2013 onwards. In addition to this, there is other nuances where 256-bit is not always better.

For example:
* The Zen and Zen+ chips provided Vector256 support but it was functionally "emulated" and so you often wouldn't see any benefit over using dual-dispatch Vector128. In addition, any APIs that did actual "cross lane" operations (that is did things like shuffling across the lower and upper 128-bits) could see performance penalties compared to just using Vector128. While this was "fixed" in Zen2 and later; there still other cases where similar can occur.
* The original CPUs that provided AVX/AVX2 support there were issues with the 256-bit instructions causing downclocking on the CPUs. This was largely mitigated by the time Skylake came out
* There are performance considerations around mixing Vector256 usage with code that is not "VEX encoding aware" which can impact throughput due the need to emit additional `vzeroupperall` instructions
* There are performance considerations around using Vector256 due to the additional ABI requirements around saving/restoring the upper 128-bits across call boundaries
* There are performance considerations on Alder Lake where Vector256 works "as expected" on the power cores but where it is less ideal to use them on the efficiency cores
* etc

Because of all of this nuance and special considerations its not really trivially to just say "use Vector<T>" by default. The option that will give users perf without them needing to consider all these implications and potential drawbacks is to just target `Vector128<T>` as the baseline. Then, if you are in a scenario where you know you consistently have enough data and where you have done the relevant profiling or disassembly checks to validate that the codegen looks good, consider also adding a `Vector256<T>` path for the additional performance increase it can provide.

Comment (stephentoub) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:54]: > The option that will give users perf without them needing to consider all these implications and potential drawbacks is to just target Vector128<T> as the baseline.

And yet the outlined guidance you provided stated if you might have data > 128 bytes, add a `Vector256<T>` code path.  Because there is a very real implication and very real drawback to "just target `Vector128<T>`", that on one of the most common architectures in the world, there is a serious gain to be had by supporting a 256-byte code path. So we're back in a world where developers _are_ going to add multiple code paths and have to know what conditions to use to guard which.

Are you prepared to simply go and delete every 256-bit code path we've added?  Developers ran microbenchmarks, saw an improvement, and that was that.  That is how it's going to be. I simply do not understand or believe the guidance "just target `Vector128<T>`" is at all realistic.

If you really believe folks shouldn't have a 256-bit code paths because ARM doesn't support it and because some older chips emulated it and because older CPUS downclocked and because there are perf considerations around ABI requirements and because of limitations on power cores, then in all seriousness we should right now go and delete all of the `Vector256<T>` cross-platform helpers that were added for .NET 7 and we should change `Vector<T>` to max out at 128-bits, period.  I do not see how in one breath we can say "look at all of this support we have for 256-bit, consider using it" and in another breath say "but you have no chance of knowing for your library whether it's worth having a 256-bit code path so just target Vector128<T> because you have no chance of always getting the 256-bit one right."

Comment (tannergooding) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:54]: We shouldn't need to rip out anything nor should we need to tell users you shouldn't use this. We do need to provide guidance here and I believe the right thing is to provide:
1. Some simple guidance that will generally work and give users a performance increase on all platforms they may want to support (even if its not the "peak" achievable performance on some architectures)
2. More complex guidance that goes into some of the nuances and considerations across various hardware, platforms, and architectures and when you may want to support the other options

That is, I'm not saying "just target `Vector128<T>`" and I'm not saying "don't target `Vector256<T>` or `Vector<T>`". What I'm trying to relay is that `Vector128<T>` is the best overall default to start with for code that doesn't need to support .NET Standard.

* `Vector128<T>` is the common SIMD baseline that guarantees you get SIMD acceleration across all the architectures we support SIMD on or will potentially support SIMD on in the future.
* `Vector128<T>` will give you the most consistent numbers across all platforms/architectures
  * This can be particularly important for developers that do not have access to or the ability to test on other hardware. It will not be perfect, but it can give a reasonable expectation of it working and what kind of gains they can likely expect
* `Vector128<T>` may not provide the absolute best performance everywhere, but it will give you the biggest relative gain over the scalar code
  *  `Vector256<T>` is only itself a gain over what `Vector128<T>` already provides and in practice it is never actually 2x more. It's generally not slower and can often be enough to warrant including it as well, but it requires additional testing and considerations
* `Vector256<T>` is never the *only* thing you should target because it is effectively x86/x64 only and would prevent SIMD acceleration on platforms like Arm64.
  * Put another way, `Vector256<T>` is an additional path you provide on top of `Vector128<T>` where the numbers show it is worthwhile
  * The numbers can change over time based on new hardware or real world usage scenarios
    * Zen/Zen+ in 2017 was an interesting case where a new micro-architecture with `Vector256<T>` support came online and didn't meet the expectations we had already assumed for our `Skylake` baseline a lot of the original microbenchmarks were taken from
    * Alder Lake from Oct 2021 is a new interesting case where it is also an entirely new microarchitecture design for x86/x64. It now has power cores (which work like we expect for Skylake/Zen2/etc) and efficiency cores (which work more like the Zen/Zen+ case) and we have not run benchmarks or taken in any numbers or considerations with how .NET runs here.
      * We already have some benchmarks that can be very noisy on Arm64 and where we believe that may be caused by differences between the big and little cores. I expect when we get some Alder Lake based machines to run benchmarks on we will see similar here as well and we will likely need to adjust our threading and other models to take into account non-homogenous CPU architectures (and this might be as simple as changing our current checks from `if (Vector256<T>.IsSupported && Length >= Vector256<T>.Count)` to also include an `&& IsPowerCore` check. But its something that needs to be considered and which we will need to be prepared to handle
* `Vector<T>` can be a good target, particularly if you also need to target .NET Standard.
  * If it is faster than the scalar code then it will likely remain faster than the scalar code regardless of its 16-bytes or 32-bytes.
  * However, the usage scenarios are somewhat limited due to it being variable sized and there are many algorithms that need things like `ExtractMostSignificantBits` to identify a matching index.
  * Likewise, due to size differences between Arm64 (always 16-bytes) and x86/x64 (32-bytes on modern hardware) users may see differing performance characteristics particularly for small inputs
    * In this PR for example; Arm64 users will see acceleration for inputs that are 32-bytes or more. x86/x64 users with AVX2 support will actually see slower code than those without for inputs that are between 32 and 64-bytes as the acceleration only kicks in at `Vector<T>.Count * 2`
    * If inputs between 32 and 64 bytes are common, then this is a de-optimization even though it likely provides additional acceleration in the case the input is larger
    * My expectation for things like strings in LoB apps is that they are small. Things that fit into an excel column or input text box (names, dates, etc). These are all things that are likely under 20 characters and so are about 40-64 bytes long.
    * My expectation is somewhat similar for certain web scenarios where you get content in chunks or packets and may need to process relatively small headers
    * There are then of course the cases where you get large strings of text. This thread is one example, PDFs, docs, and books are others. This is where things like `Vector<T>` or `Vector256<T>` have the chance to really shine

Comment (tannergooding) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:54]: I'm only trying to express that this is not a simple space and does not have a simple answer.

If users want a simple answer, then I think we can provide a decently reasonable one along the lines of what I've attempted to express above (effectively a `TL;DR`).

We can then provide additional guidance on top of that which goes in depth for users or scenarios where its worthwhile. This additional guidance would likely be what we follow in the BCL and is somewhat reliant on the fact that we are a big enterprise with lots of customers and hardware. We have the ability to extensively test (and continuously test over time) to show where gains exist, to catch regressions, and to see the differences across a range of hardware and operating systems.

Comment (davidfowl) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:54]: Do we really need to rewrite all of the vectorized code everywhere? Have you also considered ASP.NET Core in this discussion as well https://github.com/dotnet/aspnetcore/blob/3de52f5c16e5ae445abaec7810515391da7c76d5/src/Shared/ServerInfrastructure/StringUtilities.cs#L420? Are you going to review all of the places and provide some guidance?

Comment (tannergooding) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:54]: > Do we really need to rewrite all of the vectorized code everywhere?

Need to? No.
* Although in the example link you provided there are improvement in .NET 6/7 that would let you simplify some of that logic. We added support for `Vector<nint>` and `Vector<nuint>` so you could merge the `if (IntPtr.Size == 8) {} else {}` paths.

There was an ask above by Stephen for guidance on when the right choice is `Vector<T>` vs `Vector64<T>` vs `Vector128<T>` vs `Vector256<T>`.

There is a lot of nuance here and the guidance I want to write is just that, guidance. Everyone's scenarios are different and hardware is constantly evolving. It's why the architecture optimization manuals are all several hundred pages long. It's a lot like the Framework Design Guidelines where there are some `DOs` some `DO NOTs` some `CONSIDERs` and some `you can always ignore everything here if you know better or your scenario calls for it`.

There is nothing "wrong" with using `Vector<T>`, `Vector128<T>`, or `Vector256<T>`. But each come with various annotations and considerations.

The biggest consideration of `Vector<T>` is that it is variable sized. If your inputs are commonly between 16 and 32 bytes (or between 32 and 64 bytes in cases like this PR where it does `Vector<T>.Count * 2`) then you will not get SIMD acceleration when using `Vector<T>` on hardware with AVX2 support.

The biggest consideration of `Vector128<T>` is that you *could* leave some perf on the table if your inputs are commonly "big". Big here however is relative and dependent on a few factors. It could be 32-bytes and it could be 128-bytes.

The biggest consideration of `Vector256<T>` is that you will not be able to use that path on anything other than modern x86/x64 hardware.

-------------------------

If you are trying to get good performance everywhere and don't need to support .NET Standard. `Vector128<T>` is a good option. `Vector<T>` can also be a good option but comes with more considerations over `Vector128<T>`. Optionally including `Vector256<T>` can then be beneficial.

If you are trying to get the best performance everywhere, you are going to be writing complicated code and maintaining multiple paths; particularly if you need to consider both small and large inputs.

Comment (tannergooding) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:54]: I think the last thing I want to note is that any of these options is likely better than the scalar (non-SIMD) equivalent. There are exceptions of course, but generally if `Vector128<T>` is faster than the scalar; then `Vector256<T>` and `Vector<T>` will also be faster (when the path is hit).

The nuance here comes down to when `Vector<T>` vs `Vector128<T>` vs `Vector128<T> + Vector256<T>` or some other combination is the right default and/or best for the broad range of target scenarios you might want to consider.

Comment (stephentoub) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:54]: > If you are trying to get good performance everywhere and don't need to support .NET Standard. Vector128<T> is a good option.

We have many vectorized methods now. From a cursory scan, they all either use `Vector<T>` or they have both 128-bit and 256-bit code paths. Can you please list which of these implementations should ideally be changed to just use `Vector128<T>`? If that is the recommended default choice, I assume the list will include at least half of our methods. Thanks.

Comment (tannergooding) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:54]: I'm not aware of any code in the BCL where we need to or should only support `Vector128<T>`.

There are likely places where `Vector256<T>` will end up needing additional or modified checks due to the changing hardware ecosystem.

There are likely places where we would see perf benefits by adding a `Vector128<T>` path on top of `Vector<T>` to ensure that we get better perf for inputs smaller than 32-bytes (or inputs that are small enough that loop unrolling and other considerations would be better).

The BCL is special from many angles and the guidance we have for our own usage may be more complex than the default recommendation for the typical user or individual OSS maintainer looking to accelerate their app.

Comment (tannergooding) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:54]: Put another way, we are the core foundation on which a lot of other code is built and run. We have no way of statically knowing how any given app will use our own code and being the baseline the places where our code is used varies a lot.

This is different from many applications or smaller libraries out there where they can easily get away with having simpler code/guidance and still seeing good benefits. This doesn't mean that we shouldn't provide or make available the extended guidance as well nor that other libraries may not have the same or similar considerations.

But there are multiple angles and target audiences here and no one single answer.

Comment (tannergooding) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:54]: Potentially put another way in an attempt at drawing a parallel, the BCL provides `T[]`, `Span<T>`, and several other tools for working with memory.

Most apps/libraries only need to use `T[]` to be successful. They get indirect benefits by the BCL using `Span<T>`

Some apps/libraries can then get additional perf benefits by providing additional overloads which use `Span<T>` allowing for reduced allocations and handling of unmanaged memory without needing to deal with unsafe.

You can then go a step further and "hyper-optimize" by doing things like casting `int` to `uint` and then zero-extending to `nuint` or by using `MemoryMarshal.GetReference` to avoid bounds checking, etc. This is something that provides the "best performance" and is also the most complicated.

All of these options are valid and we use all of them throughout the BCL depending on what the exact circumstances and needs are.

-------------------

In the same way, it is fine for most user code to just use `Vector128<T>` and it gives them the easiest path and chance of "succeeding" (for some definition of success) on all platforms. `Vector<T>` can similarly be used to allow them to succeed and is almost as good of a choice, but there is some complicated nuance around it that makes it more confusing (at least in my own opinion) to recommend as the default scenario. One of these two is the equivalent to `T[]`.

`Vector256<T>` and providing multiple paths is somewhat like the `Span<T>` case. This can be beneficial when you know what you're doing and most code can get indirect benefits by the BCL utilizing these techniques.

Using platform specific intrinsics is a case of "hyper-optimization". You use these when you need the utmost control of the codegen and are trying to eek out every bit of performance or where the xplat helpers happen to fall short.

Comment (stephentoub) [reply src/libraries/System.Linq/src/System/Linq/Max.cs:54]: > I'm not aware of any existing SIMD code in the BCL where we need to or should only support `Vector128<T>`

Right

> The BCL is special from many angles and the guidance we have for our own usage may be more complex than the default recommendation

I do not believe the core libraries are special in this regard. If someone is reaching for `Vector`, they're already in the 1% of caring about getting really great throughput and have already demonstrated an interest in writing additional code for performance. On top of that, most libraries are in a position where they can't predict an upper bound on input lengths. 

If the guidance we have doesn't work for us, it doesn't work.

**Pull request merged**
